### ISSUE TITLE
Attestations require uploading entire payload to rekor

### ISSUE THREAD

**Issue by jonjohnsonjr** [sigstore/cosign#3599]
**Description**

Recently ran into an issue where we have an absurdly large attestation (130MB) that is rejected by rekor due to its size.

Arguably, this is a bug in rekor, but I think it's really a bug in how cosign attestations work. Both the dsse and intoto record types require uploading the entire attestation payload to rekor. If they're small enough, rekor will store the intoto attestation. AFAICT, rekor doesn't store the dsse payload at all, but it _does_ require you to upload the entire thing.

Cosign should support a form of attestation that doesn't require the entire payload be uploaded to rekor.

This is part of why I initially proposed an OCI descriptor as the payload format, since embedding `data` is an optional thing. I think sigstore ended up using dsse so that the payload type could be signed, but I can't remember the details.

Anyway, it would be great if someone could tell me I'm just holding this wrong or something, but I'd love an easy way to attest a large thing. Should I just attest an OCI descriptor that points to the large thing? Is there an existing pattern that people use for this?

It would take a lot of load off of rekor if it could just verify a hash instead of the entire payload, so maybe what we need is a form of `cosign attest` that stores and verifies a hashedrekord? Does that even make sense? If someone can hold all of this in their head, I'd love your opinions on what the right path forward would be.

**Comment by bobcallaway** [comment]
> Arguably, this is a bug in rekor, but I think it's really a bug in how cosign attestations work. Both the dsse and intoto record types require uploading the entire attestation payload to rekor. If they're small enough, rekor will store the intoto attestation. AFAICT, rekor doesn't store the dsse payload at all, but it _does_ require you to upload the entire thing.

Originally we had envisioned Rekor storing the full attestation (and implemented the `intoto` type to do just that), but after some experience and hindsight realized that it was probably better to not position Rekor to need full search capabilities (since folks want to be able to look up all log entries /attestations for arbitrary subjects). We added the `dsse` type to clean up some code issues, as well as start enforcing the pattern that Rekor wouldn't store the attestation when an entry was made with that type.

One way you could think of a DSSE envelope is that it has "embedded" signatures (similar to an RPM or an APK file), since the payload and signature are part of the same JSON document. 

It is easier for Rekor clients to upload the DSSE envelope in its entirety to verify the signature over the entire envelope VS having the client break it apart manually. It's also arguably stronger from a claimant point of view because Rekor has computed the hash directly (during the certificate's validity period) and thus 'observed' the attestation, versus trusting that the digest provided in a `hashedrekord` entry actually corresponds to a DSSE envelope that exists and was signed.

It's not ideal for Rekor to be parsing multi-hundred MB+ JSON documents just for signature verification, and that's why there's a size limit on what you can upload to the public Rekor instance.

**Comment by mrjoelkamp** [comment]
> It is easier for Rekor clients to upload the DSSE envelope in its entirety to verify the signature over the entire envelope VS having the client break it apart manually. 

The client can upload the digest of the DSSE envelope in its entirety without having to break it apart.

> It's also arguably stronger from a claimant point of view because Rekor has computed the hash directly (during the certificate's validity period) and thus 'observed' the attestation, versus trusting that the digest provided in a hashedrekord entry actually corresponds to a DSSE envelope that exists and was signed.

Verifying that the Rekor HashedRekord log entry payload matches the hash of the DSSE envelope can and should be checked by the client by computing the DSSE hash during verification. 

I have an example of this implemented for OpenPubkey attestations (experimental) here: [VerifyEntryPayload](https://github.com/openpubkey/signed-attestation/blob/5ebc1df5bc0719b1b12dd73849efb9c7f481e2ff/tl.go#L154)

**Comment by jonjohnsonjr** [comment]
What would the process be for adding support for a hashedrekord of a dsse envelope to cosign attest and cosign verify-attestation? Just send a PR, or do we need to gather some consensus around that?

**Comment by priyawadhwa** [comment]
I think sending a PR would be fine. @bobcallaway do you have any concerns?

**Comment by lkatalin** [comment]
I want to note that the problem is broader than attestations - we have this same issue with RPMs, which can be large. Uploading an RPM as a `hashedrekord` type does sidestep that problem but ignores the signature embedded in the RPM. I will try to follow the solution here (and look into DSSE envelopes) to see if it's applicable to RPMs; but any attestation-specific solution may need to be broadened to other "types".

**Comment by mrjoelkamp** [comment]
> It would slightly complicate verification, as we don't know if an entry is a hashedrekord or dsse type. We would have to canonicalize the entry as both.

Isn't this already solved since an attestation entry can be either `intoto` or `dsse` entries? AFAIK this is abstracted away in the verification path using `models.LogEntryAnon` and [RekorBundle](https://github.com/sigstore/cosign/blob/5d60a9a063a4c7c47dd63f4de43e951b55de24d8/pkg/cosign/bundle/rekor.go#L35)

> It would also mean that entries uploaded in Cosign couldn't be verified by other clients, as currently it's assumed that an attestation is uploaded as an `intoto` or `dsse` type. I'll create an issue on #clients mentioning this to see if they have any thoughts

Similar to the above, I believe this can be abstracted away using `models.LogEntryAnon` correct? And then can be unmarshalled using the generic [UnmarshalProposedEntry](https://github.com/sigstore/rekor/blob/6cdd40d56e924cf13e09b40d53a04faed5e3e8c6/pkg/generated/models/proposed_entry.go#L84) and [UnmarshalEntry](https://github.com/sigstore/rekor/blob/6cdd40d56e924cf13e09b40d53a04faed5e3e8c6/pkg/types/entries.go#L101) functions [like so](https://github.com/openpubkey/signed-attestation/blob/5ebc1df5bc0719b1b12dd73849efb9c7f481e2ff/tl.go#L205-L212).

**Comment by haydentherapper** [comment]
Yea, I noted on the Slack thread that this may not be an issue if the kind or canonicalized entry is persisted.

**Comment by BorekZnovustvoritel** [comment]
Hello, we have recently ran into this issue. Is there any progress or a workaround accessible using Cosign and Rekor CLI? We had to disable using Rekor in `cosign attest` because it rejects attestations containing SBOMs (the attestations are too large to be processed by Rekor).

**Comment by BorekZnovustvoritel** [comment]
I want to elaborate a bit on my previous comment.

Our (I am a dev from Red Hat, definitely not a spokesman) goal is to add SBOMs to the registry in a way they are signed and verifiable by anyone with the public key. We have tried using `cosign attest` with the SBOM as a payload, but some of these files are simply too big to be processed by Rekor (custom instance of v1 Rekor server).

I have thought that maybe I can attest the SBOM and only upload the HashedRekord to the Rekor server. It is possible ([here](https://github.com/BorekZnovustvoritel/Cosign-Rekor-Demo) is my short writeup with a demo), but this rekord cannot be used by `cosign verify-attestation` anyway.

@haydentherapper  I want to ask what is the roadmap for this. If it is something that you want to have addressed even in Rekor v1 and in Cosign, you can definitely describe the desired state and we will try to contribute.

Our ideas include:
- Updating both Rekor and Cosign so that a hashedRekord can be used for verifying attestations
- Only attesting the pullspec of the SBOM (what schema would we want to use for that?) and downloading it by a different command. For the convenience we would like to contribute by creating a command `cosign download blob` so that no additional tooling has to be installed and the pullspec doesn't need to be transformed into an OCI endpoint format

**Comment by Hayden-IO** [comment]
You could also work around this by specifying the DSSE's [PAE](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#:~:text=Functions%3A-,PAE,-()%20is%20the%20%22Pre) and using `cosign sign/sign-blob` rather than `cosign attest/attest-blob`. This could either be done by a separate tool or a flag on `cosign sign(-blob)`.

We could explore a client change to support uploading an payload digest, e.g. `cosign attest-blob/attest dsse.json --log-payload-digest`, though you'll need to do some investigation on how to support verification. `cosign verify-attestation(-blob)` assumes the artifact is logged as a `dsse` Rekor type, and `cosign verify(-blob)` won't verify the contents of the attestation like the subject. I suspect this will be a more complex approach than signing the PAE.

**Comment by BorekZnovustvoritel** [comment]
I have created a PR with changes that support the verification of HashedRekords in attestations. I have then verified the changes with the same approach as used in [my writeup](https://github.com/BorekZnovustvoritel/Cosign-Rekor-Demo) and the verification succeeds. I have decided not to commit the whole enhancement in a single PR as that could make it hard to review. So the whole functionality required is not included yet, but at least the final part with no workarounds available is there.

**Comment by kommendorkapten** [comment]
For a future direction I would like to see that Rekor only supports HashedRekord, and that DSSE (primarily in-toto attestations) entries are "converted" into a hashed rekord and then sign the PAE as normal. The digest of the PAE would be the hashed record. However I would like to have a clear client spec on describing this behaviour, as it will impact signing and verification of attestations, as the client will need to understand that, and for backward compatibility the existing `dsse` type in Rekor must not break. 

To unblock now, I would suggest to do what @haydentherapper is proposing; to use `sign-blob` over the PAE until we have a clear migration path defined for the entire community.

**Comment by BorekZnovustvoritel** [comment]
Thank you for your comment. However, I have some concerns about the proposed suggestion. From our customer's perspective, what would be the verification mechanism? Would it be something like this:

- User has to have both `cosign` and some other OCI client like `oras` installed
- They would have to download the blob using the other OCI client
- They would have to download the cosign bundle somehow (I guess that would have to be stored as an additional blob)
- They would have to run `cosign verify-blob...` with the bundle and the downloaded PAE blob
- If the verification passes, they then have to parse the PAE to get the SBOM

If I am missing something, please correct me (I am not that knowledgeable about OCI).

But if my assumption is correct, then I think the only think that would improve our situation is the migration to HashedRekord. Our current workaround is to ignore TLOG, which is alright as long as the keys used for signing are long-lived and single purpose. The verification of that is a single line of commands (`cosign verify-attestation --insecure-ignore-tlog` piped to `jq` and `base64`), which is much simpler than using blobs.

**Comment by kommendorkapten** [comment]
Are you using cosign V2 or V3? 

I don't remember exactly how cosign deals with blobs. But I don't think they are uploaded to any OCI regsitry, so you would get the signatures on disk (if using a V3 you would get a bundle by default). But you are right that there are some extra steps which is not great UX. 

I can look a bit more into this (hopefully tomorrow). But it seems like just switching to hashed rekord will be the least pain full option right now.

**Comment by BorekZnovustvoritel** [comment]
We used to use Cosign v2, but v3 and its bundles are more desirable for our usecase in the future.

> But it seems like just switching to hashed rekord will be the least pain full option right now.

I agree. When you find some time, you can also try to visit the PR I have opened (#4490 ), that addresses the verification of HashedRekord in Attestation using PAE. It's only one half of the problem solved (the client cannot yet create the HashedRekord-verified attestations), but it's a first step towards the best scenario.

**Comment by steiza** [comment]
I want to make sure we're on the same page before we start reviewing implementations.

As I understand it, the problem we're trying to solve is how to attest / sign SBOMs that are too large for Rekor to accept as a DSSE Rekor entry. We have to do so in a way that is backwards-compatible, and that is reasonably easy to verify. But there's a couple of layers to this problem: what should the bundle look like (and how will we verify it), and *then* how will we implement it.

Before we get to implementation, let's talk about what the bundle will look like.

1. Bundle has a hashedrekord Rekor entry and a message signature

This is what you get today if you use `cosign sign` or `sign-blob` instead of `attest` or `attest-blob`. This bundle will be small, because it will just contain the signature over the SBOM, not the SBOM itself. You would verify this with `cosign verify` or `verify-blob` and include a path to the SBOM.

2. Bundle with a hashedrekord Rekor entry and DSSE content

This bundle will be large, as it will include a base64-encoded version of whatever you are attesting (like a 100 MB+ SBOM). We don't have bundles with this combination today, so we'd need to test clients (and probably add conformance tests) to ensure that clients can understand and verify this combination.

So do I understand correctly that option 2 is what you're aiming for?

The advantages of (1) over (2) is that:
- it's supported today
- the Sigstore bundle is much smaller

The advantage of (2) over (1) is:
- you don't have to store the Sigstore bundle and the SBOM separately, since they are the same (large) document

Another variation to consider is storing the SBOM separately, and then using the [in-toto reference predicate](https://github.com/in-toto/attestation/blob/main/spec/predicates/reference.md) to link out to the SBOM (including its hash to protect integrity!) This is another way for you to use `cosign attest` / `attest-blob` today to sign SBOMs, have a small signature document, and still protect the integrity of the SBOM (although you still need to store the SBOM separately).

So before we move on to implementation, can you confirm that you want to store the SBOM in a 100 MB+ Sigstore bundle, and not instead store it separately and either sign it directly with `cosign sign` / `sign blob`, or use `cosign attest` / `attest-blob` with an in-toto reference predicate?

**Comment by BorekZnovustvoritel** [comment]
Hello, thank you for your response. I honestly don't mind any of these solutions. In my opinion the goal is just to ensure the users (verifiers) can:

- Download SBOM and verify its authenticity
- The only input for this command should be image reference (not SBOM blob reference) and a public key
- The verification and downloading should be performed in a single step to eliminate the window for adversaries to change the SBOM before the user downloads it after some first command validated some blob online

The concrete implementation (1 or 2) doesn't matter to me too much so I think it should be decided by your team. If there are drawbacks to the method 2, we shouldn't use it. From your message, option 1 seems reasonable, I just haven't found a way how to make it user friendly for the verifier with the current client version.

I have noticed you have opened a counterproposal to my implementation, which is great! I am fully open to either rework my PR (according to your guidance) or close it in favor of your solution.

**Comment by steiza** [comment]
Thanks for walking me through this!

I think you can use Sigstore libraries and primitives to achieve this use-case, but I think implementing support for this exact scenario might be out of scope for Cosign. In particular:

> The only input for this command should be image reference (not SBOM blob reference) and a public key

This would require Cosign to be opinionated about how SBOMs are stored in OCI registries, in order to reliably go from the image reference (not the SBOM reference) to at least the digest of the SBOM. We need to know the digest of the SBOM to verify, otherwise we aren't protecting against the SBOM being tampered with!

As much as possible, we want to keep OCI implementation details out of Cosign. Best practices are still being ironed out here, and we don't want to implement something in Cosign that has to keep being updated as best practices change.

This **doesn't** mean that you can't use Cosign or sigstore-go at all - just that your tooling would be responsible for figuring out the mapping from the image reference to the SBOM blob reference, and then calling `cosign verify` on that blob (or `cosign verify-attestation` if you end up using something like the [in-toto reference predicate](https://github.com/in-toto/attestation/blob/main/spec/predicates/reference.md)).

**Comment by BorekZnovustvoritel** [comment]
Sorry, that was an imprecision on my side. By image reference, I meant the one containing image digest (`quay.io/foo/bar@sha256:...`). Cosign is able to fetch image attestations, so SBOM within an attestation passes our requirements. So I think we can achieve exactly what I put into the requirements by updating how cosign attests images (mainly just the Rekor part). Without using Rekor, this exact scenario works with Cosign even today. So that's why we have chosen attestations for our usecase. If attesting supported HashedRekords, I think that's all we need for being able to use Cosign attestations for storing verifiable SBOMs.

**Comment by steiza** [comment]
We discussed this in Slack, but I'm putting the same content here so it's more durable.

The way I've been thinking about this is that the SBOM is just another object in the OCI registry. So just like you'd sign a container image (via the manifest), you could "just" use cosign to sign the SBOM manifest that's stored in the OCI registry.

For example:

Generate an image and a SBOM:
```
~/code/sigstore/sigstore-go/examples/sigstore-go-verification$ ko build .
...
```

Find the SBOM:
```
~/code/sigstore/cosign$ go run cmd/cosign/main.go tree localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a@sha256:0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9
📦 Supply Chain Security Related artifacts for an image: localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a@sha256:0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9
└── 📦 SBOMs for an image tag: localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a:sha256-0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9.sbom
   └── 🍒 sha256:a91303604ae92d028e8d32b6975f9571e4f391298221ef22df6ceead0fc5fea9
```

Make sure you have the right manifest for the SBOM:
```
~/code/sigstore/cosign$ oras manifest fetch localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a:sha256-0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9.sbom | jq
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "size": 233,
    "digest": "sha256:f7d4316c7d6925cc51a515d48c1a6606a8511b7a3011a35e7bee72fddaa9131b"
  },
  "layers": [
    {
      "mediaType": "text/spdx+json",
      "size": 3482,
      "digest": "sha256:a91303604ae92d028e8d32b6975f9571e4f391298221ef22df6ceead0fc5fea9"
    }
  ]
}
```

Sign the SBOM manifest:
```
~/code/sigstore/cosign$ go run cmd/cosign/main.go sign --key cosign2.key localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a:sha256-0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9.sbom
```

Inspect the signed SBOM manifest to find the Sigstore protobuf bundle:
```
$ go run cmd/cosign/main.go tree localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a:sha256-0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9.sbom
📦 Supply Chain Security Related artifacts for an image: localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a:sha256-0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9.sbom
└── 🔗 https://sigstore.dev/cosign/sign/v1 artifacts via OCI referrer: localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a@sha256:bc63f370ccabbf2139a6fbcd2892a33c2204c46fe8421d1048a7c5cb7767d937
   └── 🍒 sha256:2515faedb0c7e77fc2fc938172f004b33b9bfe94a7794cc8ea1445f85da65d6f
```

Look at the bundle:
```
$ oras blob fetch --output - localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a@sha256:2515faedb0c7e77fc2fc938172f004b33b9bfe94a7794cc8ea1445f85da65d6f | jq
{
  "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
...
```

Verify the SBOM with the bundle:
```
$ go run cmd/cosign/main.go verify --key cosign2.pub localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a:sha256-0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9.sbom

Verification for localhost:1338/demo/sigstore-go-verification-ce4f5bf3233ac2b951c3667b2d19da3a:sha256-0b64ad459d4a9c1281434eca87d7026e1375a2a886a66e9787db3abb847c96f9.sbom --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - The signatures were verified against the specified public key
...
```
All we really care about here is signing the OCI manifest of the SBOM (just like we sign container images by signing the manifest of that container image). We aren't using any special attestation types. We aren't including the SBOM in the attestation, so we don't need to worry about Rekor payload size.

**Comment by xnox** [comment]
many compliance and safer cryptopgrahy standards (NIST, RFC, ISO, BSI, etc) recommend to always verify full message (as in the signer & verifiers need to compute the message digest, not just passed in pre-hashed digest). Thus I really like the idea of having a "short" normal looking image manifest that is directely hashed/signed; the the bulk of the contents is a normal by-digest referenced layer.

**Comment by xnox** [comment]
Is what shown in https://github.com/sigstore/cosign/issues/3599#issuecomment-3786590559 is the deprecated and scheduled for removal SBOM attestation type - or is it just a random tag, a regular image, which happens to have .sbom suffix and a funky layer mediatype?

**Comment by codysoyland** [comment]
I'd love to come to a solution on this. There are a lot of good ideas in this thread and I think it would be good to move forward with something as SBOMs continue to get larger and this limitation is growing problem.  I think the proposals in this thread break down into two buckets:

### 1. Use `HashedRekord` for attestations
**Contributors:** [@mrjoelkamp](https://github.com/sigstore/cosign/issues/3599#issuecomment-1998633302), [@BorekZnovustvoritel](https://github.com/sigstore/cosign/issues/3599#issuecomment-3414947891), [@Hayden-IO](https://github.com/sigstore/cosign/issues/3599#issuecomment-3416354364), [@kommendorkapten](https://github.com/sigstore/cosign/issues/3599#issuecomment-3612143504)

Use `HashedRekord` — the same Rekor type used for signed blobs — for attestations instead of `dsse` or `intoto`. Rather than uploading the entire DSSE envelope to Rekor, only a hash of it would be uploaded. The approach has several variants discussed in the thread:

- **Short-term:** Add `HashedRekord` as an option for attestation uploads in `cosign attest`, so the client hashes the DSSE envelope and uploads just the digest. A [PR (#4490)](https://github.com/sigstore/cosign/pull/4490) was opened by @BorekZnovustvoritel to support verification of `HashedRekord` entries in attestations, but this logic should be moved into sigstore-go as the canonical verifier.
- **Workaround:** Compute the DSSE [Pre-Authentication Encoding (PAE)](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md) and use `cosign sign/sign-blob` instead of `cosign attest`, which already uploads a `HashedRekord`. [@Hayden-IO proposed](https://github.com/sigstore/cosign/issues/3599#issuecomment-3416354364) this as an interim solution, though [@BorekZnovustvoritel noted](https://github.com/sigstore/cosign/issues/3599#issuecomment-3612431003) it has poor UX (requires extra tools and manual PAE parsing).
- **Long-term:** [@kommendorkapten proposed](https://github.com/sigstore/cosign/issues/3599#issuecomment-3612143504) that Rekor should eventually support *only* `HashedRekord`, with DSSE/in-toto entries "converted" by hashing the PAE. This requires a clear client spec and backward compatibility plan across the sigstore ecosystem.

**Concern** (from [@Hayden-IO](https://github.com/sigstore/cosign/issues/3599#issuecomment-2007804130)): Complicates verification since clients won't know if an entry is a `hashedrekord` or `dsse` type, and entries uploaded this way couldn't be verified by other sigstore clients that assume `intoto`/`dsse` types. Though [@mrjoelkamp argued](https://github.com/sigstore/cosign/issues/3599#issuecomment-2007848429) this is already abstracted via `models.LogEntryAnon`.

**My concern** with HashedRekord is that the signed payload is pre-hashed, and therefore incompatible with ed25519 and potential future PQC standards as well as NIST's guidance.

### 2. Sign the SBOM's OCI manifest directly (treat SBOM as an OCI artifact)
**Proposed by:** [@steiza](https://github.com/sigstore/cosign/issues/3599#issuecomment-3786590559)

Don't embed the SBOM in an attestation at all. Instead, store the SBOM as a separate OCI artifact in the registry and use `cosign sign` on its manifest — the same way container images are signed. The Rekor entry only contains the OCI manifest with the hash, not the SBOM content.

**Trade-off:** Cosign won't manage the image-to-SBOM mapping; external tooling must handle that. [@steiza argued](https://github.com/sigstore/cosign/issues/3599#issuecomment-3786590559) Cosign shouldn't be opinionated about OCI storage conventions.

### My take

I'm generally in favor both of the above solutions: Using a Rekor type such as HashedRekord that doesn't require the full attestation would solve the problem in the least-friction way; this would be an opaque version bump for most people. I do however have longer-term concerns about using a pre-hashed payload. Maybe this can be worked around by adding a second stronger signature to the DSSE that clients would be required to also verify, but that would require a larger discussion.

The second option of uploading the SBOM as an independent registry object and signing its manifest feels more "correct", but this is not a drop-in solution for people who are already committed to using SBOM attestations; the attestation subject becomes an OCI manifest, thus it couples the format with OCI formats and the subject-to-SBOM chain is more complicated.

To offer a slightly more nuanced version of the SBOM manifest suggestion: we can also utilize the OCI 1.1 referrers API to tie things together more concretely. The SBOM's manifest can declare the signed image to be the `subject`, which means the referrers API would return that SBOM in its list of referrers. Then, the cosign signature manifest would declare the SBOM as its subject, completing the chain. For example, the following `oras discover` call will recursively identify the SBOM associated with the image and its detached signature:

```
❯ oras discover --plain-http localhost:5050/demo/busybox:latest
localhost:5050/demo/busybox@sha256:1487d0af5f52b4ba31c7e465126ee2123fe3f2305d638e7827681e7cf6c83d5e
└── application/spdx+json
    └── sha256:4041314ffb178f343f306000bea36f948b0c2b145aef538e836c65d34e1b9374
        ├── [annotations]
        │   └── org.opencontainers.image.created: "2026-05-05T19:45:20Z"
        └── application/vnd.dev.sigstore.bundle.v0.3+json
            └── sha256:d0b4828c31b55dcc36d9e18e32b34661adc39c53a4dbcc6fd40ce3f7f8b047c7
                └── [annotations]
                    ├── org.opencontainers.image.created: "2026-05-05T19:46:34Z"
                    ├── dev.sigstore.bundle.content: dsse-envelope
                    └── dev.sigstore.bundle.predicateType: https://sigstore.dev/cosign/sign/v1
```

This also means that `oras cp --recursive` will cleanly copy the image, sbom, and signature in one call.

I actually think there is room for both of these proposals in Cosign: provide an option in the signers/verifiers to use hashedrekord entries as the low-hanging fruit option that unblocks large attestation bundles, and also add some kind of mode to recursively sign/verify the associations between signed SBOMs and their subject manifests. These can both be flags on the existing subcommands in cosign.
